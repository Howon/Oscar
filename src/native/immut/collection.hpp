#ifndef __COLLECTION_HPP__
#define __COLLECTION_HPP__

#include <iostream>
#include <sstream>
#include <atomic>
#include <memory>
#include <stdexcept>
#include <cassert>
#include <functional>
#include <initializer_list>

#include <stdlib.h>
#include <time.h>

using namespace std;

namespace immut {
    enum Color { R, B };
    enum Type {L, S, M};

    template <typename T>
    class collection {
        Type type;

    public:
        Type getType() const { return type; }

        virtual bool isEmpty() const = 0;

        virtual bool contains(T x) const = 0;
    };
}

#endif
