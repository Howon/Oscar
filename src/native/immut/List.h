#ifndef __LIST_H__
#define __LIST_H__

#include <iostream>
#include <cassert>
#include <memory>
#include <functional>
#include <initializer_list>
#include <iterator>

#include "collection.h"

using namespace std;

namespace immut {
  template <typename T>
  class list : public collection<T> {
      Type type = L;

      struct Item {
          Item(T v, shared_ptr<const Item> tail)
              : _val(v), _next(move(tail))
          {}

          explicit Item(T v) : _val(v) {}

          T _val;

          shared_ptr<const Item> _next;
      };

      friend Item;

      explicit list(shared_ptr<const Item> items) : _head(move(items)) {}

      shared_ptr<const Item> _head;

  public:
      list(T v, list const &tail) : _head(make_shared<Item>(v, tail._head)) {}

      list(list const &head, T v) : _head(make_shared<Item>(head._head, v)) {}

      explicit list(T v) : _head(make_shared<Item>(v)) {}

      list(initializer_list<T> init)
      {
          for (auto it = rbegin(init); it != rend(init); ++it) {
              _head = make_shared<Item>(*it, _head);
          }
      }

      virtual bool isEmpty() const { return !_head; } // conversion to bool

      virtual T front() const
      {
          assert(!isEmpty());

          return _head->_val;
      }

      list pop_front() const
      {
          assert(!isEmpty());

          return list(_head->_next);
      }

      list push_front(T v) const
      {
          return list(v, *this);
      }

      list push_back(T v) const
      {
          return list(*this, v);
      }

      list take(int n)
      {
          if (n <= 0 || isEmpty())
            return list();

          return pop_front().take(n - 1).push_front(front());
      }

      list insertedAt(int i, T v) const
      {
          if (i == 0)
              return push_front(v);
          else {
              assert(!isEmpty());

              return list<T>(front(), pop_front().insertedAt(i - 1, v));
          }
      }

      list removed(T v) const
      {
          if (isEmpty())
            return list();
          if (v == front())
              return pop_front();

          return list(front(), pop_front().removed(v));
      }

      bool contains(T v) const
      {
          if (isEmpty())
            return false;

          if (v == front())
            return true;

          return pop_front().contains(v);
      };

      template <typename F>
      void forEach(const F f) const
      {
        cout << " hererr" << endl;
          Item const * it = _head.get();
    cout << it->_val << endl;
          while (it != nullptr) {
              f(it->_val);
              it = it->_next.get();
          }
      }

      virtual ostream &operator<<(ostream &os)
      {
          os << "[ ";

          forEach([&os](T v) {
              os << v << " ";
          });

          os << "]";

          return os;
      }
  };
}

#endif
