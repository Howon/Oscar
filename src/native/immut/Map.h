#ifndef __MAP_H__
#define __MAP_H__

#include <vector>
#include "collection.h"

using namespace std;

namespace immut {
    template <typename K, typename V>
    class map : public collection<K>
    {
        Type type = M;

        struct KVNode
        {
            KVNode(Color c, shared_ptr<const KVNode> const &lft,
                K key, V val, shared_ptr<const KVNode> const & rgt)
            {
                _hash = rand();
                _c = c;
                _lft = lft;
                _rgt = rgt;
                _key = key;
                _val = val;
            }

            int _hash;

            Color _c;
            K _key;
            V _val;

            shared_ptr<const KVNode> _lft;
            shared_ptr<const KVNode> _rgt;
        };

        explicit map(shared_ptr<const KVNode> const &node) {
            _root = node;
        }

        shared_ptr<const KVNode> _root;

        Color rootColor() const
        {
            assert(!isEmpty());

            return _root->_c;
        }

        map paint(Color c) const
        {
            assert(!this->isEmpty());

            return map(c, left(), this->rootKey(), this->rootValue(), right());
        }

        int countB() const
        {
            if (this->isEmpty())
                return 0;

            int lft = left().countB();

            assert(lft == right().countB());

            return (this->rootColor() == B)? 1 + lft: lft;
        }

        map left() const
        {
            assert(!this->isEmpty());

            return map(_root->_lft);
        }

        map right() const
        {
            assert(!this->isEmpty());

            return map(_root->_rgt);
        }

        bool doubledLeft() const
        {
            return !this->isEmpty() && this->rootColor() == R &&
                !this->left().isEmpty() && this->left().rootColor() == R;
        }

        bool doubledRight() const
        {
            return !this->isEmpty() && this->rootColor() == R &&
                !this->right().isEmpty() && this->right().rootColor() == R;
        }

        map balance(const map &lft, K x, V v, const map &rgt) const
        {
            if (lft.doubledLeft()) {
                return map(R, lft.left().paint(B), lft.rootKey(),
                    lft.rootValue(), map(B, lft.right(), x, v, rgt));
            } else if (lft.doubledRight()) {
                return map(R, map(B, lft.left(), lft.rootKey(), lft.rootValue(),
                    lft.right().left()), lft.right().rootKey(),
                    lft.right().rootValue(), map(B, lft.right().right(),
                    x, v, rgt));
            } else if (rgt.doubledLeft()) {
                return map(R, map(B, lft, x, v, rgt.left().left()),
                    rgt.left().rootKey(), rgt.left().rootValue(),
                    map(B, rgt.left().right(), rgt.rootKey(),
                    rgt.rootValue(), rgt.right()));
            } else if (rgt.doubledRight()) {
                return map(R, map(B, lft, x, v, rgt.left()), rgt.rootKey(),
                    rgt.rootValue(), rgt.right().paint(B));
            } else
                return map(B, lft, x, v, rgt);
        }

        virtual void assert1() const
        {
            if (!this->isEmpty()) {
                auto lft = left();
                auto rgt = right();

                if (this->rootColor() == R) {
                    assert(lft.isEmpty() || lft.rootColor() == B);
                    assert(rgt.isEmpty() || rgt.rootColor() == B);
                }

                lft.assert1();
                rgt.assert1();
            }
        }

    public:
        map() {}

        map(Color c, const map &lft, K key, V val, const map &rgt) :
            _root(make_shared<const KVNode>(c, lft._root, key, val, rgt._root))
        {
            assert(lft.isEmpty() || lft.rootKey() < key);
            assert(rgt.isEmpty() || key < rgt.rootKey());
        }

        map(initializer_list<pair<K, V> > initMap)
        {
            map<K, V> m;

            for (auto it = initMap.begin(); it != initMap.end(); ++it)
                m = m.inserted(it->first, it->second);

            this->_root = m.get_root();
        }

        shared_ptr<const KVNode> get_root() { return _root; }

        K rootKey() const
        {
            assert(!this->isEmpty());

            return _root->_key;
        }

        V rootValue() const
        {
            assert(!this->isEmpty());

            return _root->_val;
        }

        bool isEmpty() const { return !_root; }

        template <typename F>
        void forEach(F f)  const
        {
            static_assert(is_convertible<F, function<void(K, V)>>::value,
                 "ForEach requires a function type void(K, V)");

            if (!this->isEmpty()) {
                this->left().forEach(f);

                f(this->rootKey(), rootValue());

                this->right().forEach(f);
            }
        }

        map insert(K x, V v) const
        {
            this->assert1();

            if (this->isEmpty())
                return map(R, map(), x, v, map());

            K y = this->rootKey();
            V yv = this->rootValue();

            if (x == y)
                return map(B, this->left(), x, yv, this->right());

            Color c = this->rootColor();

            if (this->rootColor() == B) {
                if (x < y) {
                    return balance(this->left().insert(x, v), y, yv,
                        this->right());
                }

                return balance(this->left(), y, yv, this->right().insert(x, v));
            } else {
                if (x < y) {
                    return map(c, this->left().insert(x, v),
                        y, yv, this->right());
                }

                return map(c, this->left(), y, yv, this->right().insert(x, v));
            }
        }

        map  inserted(K x, V v) const
        {
            auto t = insert(x, v);

            return map(B, t.left(), t.rootKey(), t.rootValue(), t.right());
        }

        virtual bool contains(K x) const
        {
            if (this->isEmpty())
                return false;

            K y = this->rootKey();

            if (x < y)
                return left().contains(x);
            else if (y < x)
                return right().contains(x);
            else
                return true;
        }

        V find(K key) const
        {
            if (this->isEmpty())
                throw out_of_range("Key not found");

            K y = this->rootKey();

            if (key < y)
                return left().find(key);
            else if (y < key)
                return right().find(key);
            else
                return rootValue();
        }

        bool operator==(const map &rhs)
        {
            forEach([&rhs](K k, V v) {
                if (!rhs.contains(k) || rhs.find(k) != v)
                    return false;
            });

            return true;
        }

        bool operator > (const map &rhs)
        {
            return get_root().get()->_hash > rhs.get_root().get()->_hash;
        }

        bool operator < (const map &rhs)
        {
            return get_root().get()->_hash < rhs.get_root().get()->_hash;
        }

        friend ostream &operator<<(ostream &os, const map &m)
        {
            os << "[ ";

            m.forEach([&os](K k, V v) {
                os << "(" << k << " -> " << v << ") ";
            });

            os << "]";

            return os;
        }
    };
}

#endif
