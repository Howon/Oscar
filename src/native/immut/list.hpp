#ifndef __LIST_HPP__
#define __LIST_HPP__

#include "collection.hpp"

using namespace std;

namespace immut {
    template <typename T>
    class list : public collection<T> {
        Type type = L;

        struct Item {
            Item(T v, shared_ptr<const Item> tail) :
                _hash(rand()), _val(v), _next(move(tail)) {}

            explicit Item(T v) : _val(v) {}

            int _hash;
            T _val;

            shared_ptr<const Item> _next;
        };

        friend Item;

        shared_ptr<const Item> _head;

        explicit list(shared_ptr<const Item> items) : _head(move(items)) {}

    public:
        list() {}

        list(T v, list const &tail) :
            _head(make_shared<Item>(v, tail._head)) {}

        explicit list(T v) : _head(make_shared<Item>(v)) {}

        list(initializer_list<T> init)
        {
            for (auto it = rbegin(init); it != rend(init); ++it) {
                _head = make_shared<Item>(*it, _head);
            }
        }

        shared_ptr<const Item> get_root() const { return _head; }

        virtual bool isEmpty() const { return !_head; }

        virtual T front() const
        {
            assert(!isEmpty());

            return _head->_val;
        }

        list pop_front() const
        {
            assert(!isEmpty());

            return list(_head->_next);
        }

        list push_front(T v) const { return list(v, *this); }

        list take(int n)
        {
            if (n <= 0 || isEmpty())
              return list();

            return pop_front().take(n - 1).push_front(front());
        }

        list insertedAt(int i, T v) const
        {
            if (i == 0)
                return push_front(v);
            else {
                assert(!isEmpty());

                return list<T>(front(), pop_front().insertedAt(i - 1, v));
            }
        }

        list removeAt(int i, int j) const
        {
            if (isEmpty())
              return list();

            if (i == j)
                return pop_front();

            return list(front(), pop_front().removeAt(i + 1, j));
        }

        list removed(T v) const
        {
            if (isEmpty())
              return list();

            if (v == front())
                return pop_front();

            return list(front(), pop_front().removed(v));
        }

        bool contains(T v) const
        {
            if (isEmpty())
              return false;

            if (v == front())
              return true;

            return pop_front().contains(v);
        };

        template <typename F>
        void forEach(const F f) const
        {
            Item const *it = get_root().get();

            while (it != nullptr) {
                f(it->_val);

                it = it->_next.get();
            }
        }

        bool operator==(const list &rhs)
        {
            Item const *it1 = get_root().get();
            Item const *it2 = rhs.get_root().get();

            while (!(it1 == nullptr || it2 == nullptr)) {
                if (it1->_val != it2->_val)
                    return false;

                it1 = it1->_next.get();
                it2 = it2->_next.get();
            }

            if ((it1 == nullptr && it2 != nullptr) ||
                  (it1 != nullptr && it2 == nullptr))
              return false;

            return true;
        }

        bool operator > (const list &rhs)
        {
            return get_root().get()->_hash > rhs.get_root().get()->_hash;
        }

        bool operator < (const list &rhs)
        {
            return get_root().get()->_hash < rhs.get_root().get()->_hash;
        }

        const T operator[](int idx) const
        {
            const Item *it1 = get_root().get();

            while (idx--) {
                if (it1 == nullptr)
                    throw invalid_argument("Index out of bound");

                it1 = it1->_next.get();
            }

            return it1->_val;
        }

        string str() const
        {
            stringstream os;

            os << "[ ";

            forEach([&os](T v) {
                os << v << " ";
            });

            os << "]";

            return os.str();
        }

        friend ostream &operator<<(ostream &os, const list &t)
        {
            return os << t.str();
        }
    };
}

#endif
