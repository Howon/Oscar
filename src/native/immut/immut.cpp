#include "list.h"
#include "map.h"
#include "set.h"

#include <vector>
#include <string>

template <typename T, typename F>
void ForEach(const F f, const immut::list<T> &t)
{
    t.forEach(f);
}

template <typename T, typename F>
void ForEach(const F f, const immut::set<T> &t)
{
    t.forEach(f);
}

template <typename K, typename V, typename F>
void ForEach(const F f, const immut::map<K, V> &t)
{
    t.forEach(f);
}

template<class T, class U, class F>
U FoldLeft(F f, U acc, immut::list<T> lst)
{
    static_assert(is_convertible<F, function<U(U, T)>>::value,
        "FoldLeft requires a function type U(U, T)");

    ForEach([&f, &acc](const T &v){
        acc = f(acc, v);
    }, lst);

    return acc;
}

template<class T>
immut::list<T> Reverse(immut::list<T> const &lst)
{
    return FoldLeft([](immut::list<T> const &acc, T v) {
        return immut::list<T>(v, acc);
    }, immut::list<T>(), lst);
}

template <typename T>
size_t Size(immut::list<T> l)
{
    return FoldLeft([](size_t acc, T t) -> size_t {
        return acc + 1;
    }, 0, l);
}

template <typename T>
size_t Size(immut::set<T> s)
{
    size_t size = 0;

    ForEach([&size](T t) {
        size++;
    }, s);

    return size;
}

template <typename K, typename V>
size_t Size(immut::map<K, V> m)
{
    size_t size = 0;

    ForEach([&size](K k, V v) {
        size++;
    }, m);

    return size;
}


template <typename T>
immut::set<T> Put(T x, immut::set<T> s)
{
    return s.inserted(x);
}

template <typename K, typename V>
immut::map<K, V> Put(K k, V v, immut::map<K, V> m)
{
    return m.inserted(k, v);
}

template <typename T, typename F>
immut::list<T> Filter(F f, immut::list<T> t)
{
    static_assert(is_convertible<F, function<bool(T)>>::value,
        "Filter requires a function type bool(T)");

    immut::list<T> res;

    ForEach([&f, &res, &t](T const &v){
        if (f(v))
            res = res.push_front(v);
    }, t);

    return Reverse<T>(res);
}

template <typename T, typename F>
immut::set<T> Filter(F f, immut::set<T> t)
{
    static_assert(is_convertible<F, function<bool(T)>>::value,
        "Filter requires a function type bool(T)");

    immut::set<T> res;

    ForEach([&f, &res, &t](T const &v){
        if (f(v))
            res = Put(v, res.inserted(v));
    }, t);

    return res;
}

template <typename K, typename V, typename F>
immut::map<K, V> Filter(F f, immut::map<K, V> t)
{
    static_assert(is_convertible<F, function<bool(K, V)>>::value,
        "Filter requires a function type bool(K, V)");

    immut::map<K, V> res;

    ForEach([&f, &res, &t](const K &k, const V &v){
        if (f(k, v))
            res = Put(k, v, res);
    }, t);

    return res;
}

template <typename T>
bool Contains(T t, immut::list<T> l)
{
    return l.contains(t);
}

template <typename T>
bool Contains(T t, immut::set<T> s)
{
    return s.contains(t);
}

template <typename K, typename V>
bool Contains(K k, immut::map<K, V> m)
{
    return m.contains(k);
}

template<class T, class F>
auto Map(F f, immut::list<T> t) -> immut::list<decltype(f(t.front()))>
{
    using U = decltype(f(t.front()));

    static_assert(is_convertible<F, function<U(T)>>::value,
        "Map requires a function type U(T)");

    immut::list<U> res;

    t.forEach([&f, &res](const T &v) { res = res.push_front(f(v)); });

    return Reverse<U>(res);
}

template<class T, class F>
auto Map(F f, immut::set<T> t) -> immut::set<decltype(f(t.front()))>
{
    using U = decltype(f(t.front()));

    static_assert(is_convertible<F, function<U(T)>>::value,
        "Met requires a function type U(T)");

    immut::set<U> res;

    t.forEach([&f, &res](const T &v) { res = Put<U>(f(v), res); });

    return res;
}

template <typename K, typename V, typename F>
auto Map(F f, immut::map<K, V> t) ->
    immut::map<K, decltype(f(t.rootKey(), t.rootValue()))>
{
    using U = decltype(f(t.rootKey(), t.rootValue()));

    static_assert(is_convertible<F, function<U(K, V)>>::value,
        "Map requires a function type U(K, V)");

    immut::map<K, U> res;

    t.forEach([&f, &res](K const &k, V const &v) {
        res = Put(k, f(k, v), res);
    });

    return res;
}

template <typename T, typename U, typename F>
U Reduce(F f, immut::list<T> lst)
{
    static_assert(is_convertible<F, function<U(T, T)>>::value,
        "Reduce requires a function type U(T, T)");
    static_assert(!lst.isEmpty(),
        "Reduce requires a nonempty collection");

    auto next = lst.popped_front;

    return FoldLeft(f, f(lst.front(), next.front()), next.popped_front());
}

template <typename T>
immut::list<T> MergeFront(immut::list<T> const &a, immut::list<T> const &b)
{
    if (a.isEmpty())
        return b;

    return immut::list<T>(a.front(), MergeFront(a.pop_front(), b));
}

template <typename T>
immut::list<T> MergeBack(immut::list<T> const &a, immut::list<T> const &b)
{
    return MergeFront(b, a);
}

template <typename T>
immut::list<T> PopFront(immut::list<T> const l)
{
    return l.pop_front();
}

template <typename T>
immut::list<T> PopBack(immut::list<T> const l)
{
    size_t size = Size(l) - 1;

    return l.removeAt(0, size);
}

template <typename T>
immut::list<T> PushFront(T t, immut::list<T> const l)
{
    return l.push_front(t);
}

template <typename T>
immut::list<T> PushBack(T t, immut::list<T> const l)
{
    auto r1 = Reverse(l);
    auto r2 = r1.push_front(t);

    return Reverse(r2);
}

template <typename T>
immut::set<T> Union(immut::set<T> const & a, immut::set<T> const & b)
{
    immut::set<T> res = a;

    ForEach([&res, &a](const T &v){
        if (!a.contains(v))
            res = res.inserted(v);
    }, b);

    return res;
}

template<class T>
immut::set<T> Diff(immut::set<T> const & a, immut::set<T> const & b)
{
    immut::set<T> res;

    ForEach([&res, &a](const T &v){
        if (!a.contains(v))
            res = res.inserted(v);
    }, b);

    return res;
}

template<class T>
immut::set<T> Intersection(immut::set<T> const & a, immut::set<T> const & b)
{
    immut::set<T> res;

    ForEach([&res, &a](const T &v){
        if (a.contains(v))
            res = res.inserted(v);
    }, b);

    return res;
}

template<class T>
bool SubSet(immut::set<T> const & a, immut::set<T> const & b)
{
    bool is_subset = true;

    ForEach([&is_subset, &b](const T &v) {
        if (!b.contains(v))
            is_subset = false;
    }, a);

    return is_subset;
}

/*
void test_list(void)
{
    immut::list<int> lst = { 1, 2, 3, 4, 5, 6, 7, 8, 9};

    cout << lst << endl;

    auto mapped = Map([](int x) -> immut::list<double> {
        return immut::list<double>{ x + 0.5 };
    }, lst);

    cout << mapped << endl;

    auto fold = FoldLeft([](double acc, int x) -> double {
        return acc * x;
    }, 10.0, lst);

    cout << fold << endl;

    auto filter = Filter([](int x) -> bool {
        return x % 2 == 0;
    }, lst);

    cout << filter << endl;

    auto mergeF = MergeFront(lst, filter);
    auto mergeB = MergeBack(lst, filter);

    cout << mergeF << endl;
    cout << mergeB << endl;

    auto pushF = PushFront(0, lst);
    auto pushB = PushBack(0, lst);

    auto popF = PopFront(lst);
    auto popB = PopBack(lst);

    cout << pushF << endl;
    cout << pushB << endl;

    cout << popF << endl;
    cout << popB << endl;

    cout << Contains(1, lst) << endl;
    cout << Contains(11, lst) << endl;
    cout << Size(lst) << endl;
    cout << lst << endl << endl;
}

void test_set(void)
{
    immut::set<int> set = {1, 2, 3, 4, 5, 6, 7, 8, 9};
    immut::set<int> set2 = {4, 5, 6, 7, 8, 9, 9, 10, 11, 12};
    immut::set<int> set3 = {1, 2, 3, 4, 5, 6, 9};

    cout << set << endl;
    cout << set2 << endl;

    auto mapped = Map([](int x) -> immut::list<double> {
        return immut::list<double>{ x + 0.5 };
    }, set);

    cout << mapped << endl;

    auto filter = Filter([](int x) -> bool {
        return x % 2 == 0;
    }, set);

    cout << filter << endl;

    auto uni = Union(set, set2);
    auto diff = Diff(set, set2);
    auto intersect = Intersection(set, set2);

    cout << uni << endl;
    cout << diff << endl;
    cout << intersect << endl;
    cout << SubSet(set, set2) << endl;
    cout << SubSet(set, set3) << endl;
    cout << SubSet(set3, set) << endl;
    cout << Put(12, set) << endl;
    cout << Size(set) << endl;
    cout << set << endl << endl;
}

void test_map(void)
{
    immut::map<string, int> map = { {"a", 1}, {"b", 2}, {"c", 3}, {"d", 4} };

    auto mapped = Map([](string s, int x) -> immut::list<double> {
        return immut::list<double> { x + 0.5, 4.5 };
    }, map);

    cout << mapped << endl;

    auto filter = Filter([](string s, int x) -> bool {
        return x % 2 == 0;
    }, map);

    cout << filter << endl;
    cout << Contains(string("b"), map) << endl;
    cout << Contains(string("z"), map) << endl;
    cout << Put(string("h"), 12, map) << endl;
    cout << Size(map) << endl;
    cout << map << endl;
}

int main(void)
{
    test_list();
    test_set();
    test_map();

    return 0;
}
*/
