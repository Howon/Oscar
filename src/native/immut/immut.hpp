#ifndef __IMMUT_HPP__
#define __IMMUT_HPP__

#include "list.hpp"
#include "map.hpp"
#include "set.hpp"

#include <iomanip>
#include <string>

template <typename T, typename F>
void ForEach(const F f, const immut::list<T> &t)
{
    t.forEach(f);
}

template <typename T, typename F>
void ForEach(const F f, const immut::set<T> &t)
{
    t.forEach(f);
}

template <typename K, typename V, typename F>
void ForEach(const F f, const immut::map<K, V> &t)
{
    t.forEach(f);
}

template<class T, class U, class F>
U FoldLeft(F f, U acc, immut::list<T> lst)
{
    static_assert(is_convertible<F, function<U(U, T)>>::value,
        "FoldLeft requires a function type U(U, T)");

    ForEach([&f, &acc](const T &v){
        acc = f(acc, v);
    }, lst);

    return acc;
}

template<class T>
immut::list<T> Reverse(immut::list<T> const &lst)
{
    return FoldLeft([](immut::list<T> const &acc, T v) {
        return immut::list<T>(v, acc);
    }, immut::list<T>(), lst);
}

int Size(string s)
{
    return s.size();
}

template <typename T>
int Size(immut::list<T> l)
{
    return FoldLeft([](size_t acc, T t) -> size_t {
        return acc + 1;
    }, 0, l);
}

template <typename T>
int Size(immut::set<T> s)
{
    size_t size = 0;

    ForEach([&size](T t) {
        size++;
    }, s);

    return size;
}

template <typename K, typename V>
int Size(immut::map<K, V> m)
{
    size_t size = 0;

    ForEach([&size](K k, V v) {
        size++;
    }, m);

    return size;
}


template <typename T>
immut::set<T> Put(T x, immut::set<T> s)
{
    return s.inserted(x);
}

template <typename K, typename V>
immut::map<K, V> Put(K k, V v, immut::map<K, V> m)
{
    return m.inserted(k, v);
}

template <typename T, typename F>
immut::list<T> Filter(F f, immut::list<T> t)
{
    static_assert(is_convertible<F, function<bool(T)>>::value,
        "Filter requires a function type bool(T)");

    immut::list<T> res;

    ForEach([&f, &res, &t](T const &v){
        if (f(v))
            res = res.push_front(v);
    }, t);

    return Reverse<T>(res);
}

template <typename T, typename F>
immut::set<T> Filter(F f, immut::set<T> t)
{
    static_assert(is_convertible<F, function<bool(T)>>::value,
        "Filter requires a function type bool(T)");

    immut::set<T> res;

    ForEach([&f, &res, &t](T const &v){
        if (f(v))
            res = Put(v, res.inserted(v));
    }, t);

    return res;
}

template <typename K, typename V, typename F>
immut::map<K, V> Filter(F f, immut::map<K, V> t)
{
    static_assert(is_convertible<F, function<bool(K, V)>>::value,
        "Filter requires a function type bool(K, V)");

    immut::map<K, V> res;

    ForEach([&f, &res, &t](const K &k, const V &v){
        if (f(k, v))
            res = Put(k, v, res);
    }, t);

    return res;
}

template <typename T>
bool Contains(T t, immut::list<T> l)
{
    return l.contains(t);
}

template <typename T>
bool Contains(T t, immut::set<T> s)
{
    return s.contains(t);
}

template <typename K, typename V>
bool Contains(K k, immut::map<K, V> m)
{
    return m.contains(k);
}

template<class T, class F>
auto Map(F f, immut::list<T> t) -> immut::list<decltype(f(t.front()))>
{
    using U = decltype(f(t.front()));

    static_assert(is_convertible<F, function<U(T)>>::value,
        "Map requires a function type U(T)");

    immut::list<U> res;

    t.forEach([&f, &res](const T &v) { res = res.push_front(f(v)); });

    return Reverse<U>(res);
}

template<class T, class F>
auto Map(F f, immut::set<T> t) -> immut::set<decltype(f(t.front()))>
{
    using U = decltype(f(t.front()));

    static_assert(is_convertible<F, function<U(T)>>::value,
        "Met requires a function type U(T)");

    immut::set<U> res;

    t.forEach([&f, &res](const T &v) { res = Put<U>(f(v), res); });

    return res;
}

template <typename K, typename V, typename F>
auto Map(F f, immut::map<K, V> t) ->
    immut::map<K, decltype(f(t.rootKey(), t.rootValue()))>
{
    using U = decltype(f(t.rootKey(), t.rootValue()));

    static_assert(is_convertible<F, function<U(K, V)>>::value,
        "Map requires a function type U(K, V)");

    immut::map<K, U> res;

    t.forEach([&f, &res](K const &k, V const &v) {
        res = Put(k, f(k, v), res);
    });

    return res;
}

template <typename T, typename U, typename F>
U Reduce(F f, immut::list<T> lst)
{
    static_assert(is_convertible<F, function<U(T, T)>>::value,
        "Reduce requires a function type U(T, T)");
    static_assert(!lst.isEmpty(),
        "Reduce requires a nonempty collection");

    auto next = lst.popped_front;

    return FoldLeft(f, f(lst.front(), next.front()), next.popped_front());
}

template <typename T>
immut::list<T> MergeFront(immut::list<T> const &a, immut::list<T> const &b)
{
    if (a.isEmpty())
        return b;

    return immut::list<T>(a.front(), MergeFront(a.pop_front(), b));
}

template <typename T>
immut::list<T> MergeBack(immut::list<T> const &a, immut::list<T> const &b)
{
    return MergeFront(b, a);
}

template <typename T>
immut::list<T> PopFront(immut::list<T> const l)
{
    return l.pop_front();
}

template <typename T>
immut::list<T> PopBack(immut::list<T> const l)
{
    size_t size = Size(l) - 1;

    return l.removeAt(0, size);
}

template <typename T>
immut::list<T> Prepend(T t, immut::list<T> const l)
{
    return l.push_front(t);
}

template <typename T>
immut::list<T> Append(T t, immut::list<T> const l)
{
    auto r1 = Reverse(l);
    auto r2 = r1.push_front(t);

    return Reverse(r2);
}

template <typename T>
immut::set<T> Union(immut::set<T> const & a, immut::set<T> const & b)
{
    immut::set<T> res = a;

    ForEach([&res, &a](const T &v){
        if (!a.contains(v))
            res = res.inserted(v);
    }, b);

    return res;
}

template<class T>
immut::set<T> Diff(immut::set<T> const & a, immut::set<T> const & b)
{
    immut::set<T> res;

    ForEach([&res, &b](const T &v){
        if (!b.contains(v))
            res = res.inserted(v);
    }, a);

    return res;
}

template<class T>
immut::set<T> Intersection(immut::set<T> const & a, immut::set<T> const & b)
{
    immut::set<T> res;

    ForEach([&res, &a](const T &v){
        if (a.contains(v))
            res = res.inserted(v);
    }, b);

    return res;
}

template<class T>
bool SubSet(immut::set<T> const & a, immut::set<T> const & b)
{
    bool is_subset = true;

    ForEach([&is_subset, &b](const T &v) {
        if (!b.contains(v))
            is_subset = false;
    }, a);

    return is_subset;
}

int AsInt(double d) { return int(d); }

double AsDouble(int i) { return double(i); }

std::string AsString(int i) { return std::to_string(i); }

std::string AsString(char c) { return std::to_string(c); }

std::string AsString(bool b) { return std::to_string(b); }

std::string AsString(double d) { return std::to_string(d); }

template <typename T>
std::string AsString(immut::list<T> l) { return l.str(); }

template <typename T>
std::string AsString(immut::set<T> s) { return s.str(); }

template <typename K, typename V>
std::string AsString(immut::map<K, V> m) { return m.str(); }

void Println(int i) { std::cout << i << std::endl; }

void Println(char c) { std::cout << c << std::endl; }

void Println(bool b) {
    std::cout << (b == true ? "true" : "false") << std::endl;
}

void Println(double d) {
    std::cout << fixed << setprecision(10);
    std::cout << d << std::endl;
}

void Println(string s) { std::cout << s << std::endl; }

template <typename T>
void Println(immut::list<T> l) { std::cout << l << std::endl; }

template <typename T>
void Println(immut::set<T> s) { std::cout << s << std::endl; }

template <typename K, typename V>
void Println(immut::map<K, V> m) { std::cout << m << std::endl; }

#endif
