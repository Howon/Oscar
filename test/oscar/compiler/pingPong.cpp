#include <oscar/immut.hpp>

int main () 
{
auto pong=actor<a_Pong>();
auto ping=actor<a_Ping>(pong);
m_setMaxTurns(3) |> ping;
m_startMsg() |> ping;
return 0;
}

