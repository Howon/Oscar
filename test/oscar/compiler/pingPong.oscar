message setPing(ping: actor);
message setPong(pong: actor);
message setMaxTurns(max: int);
message startMsg();
message stopMsg();
message pingMsg();
message pongMsg();

actor Ping {
  mut int count = 0;
  mut int maxTurns = 5;
  mut actor pong;

  def incrementAndPrint() => unit = {
    count = count + 1;
    println("ping");
  }

  receive {
    | setPong(myPong: actor) => {
        pong = myPong;
      }
    | setMaxTurns(max: int) => {
        maxTurns = max;
      }
    | startMsg() => {
        incrementAndPrint();
        pingMsg |> pong;
      }
    | pongMsg() => {
        incrementAndPrint();

        if (count > maxTurns) {
            println("ping stopped");
            stopMsg |> pong;
        } else {
            pingMsg |> pong;
        }
      }
  }
}

actor Pong {
  mut actor ping;

  receive {
    | setPing(myPing: actor) => {
          ping = myPing;
      }
    | pingMsg() => {
        println("  pong");
        pongMsg |> ping
      }
    | stopMsg() => {
        println("pong stopped");
      }
  }
}


def main() => unit = {
  Ping ping = spawn Ping();
  Pong pong = spawn Pong();

  setPong(pong) |> ping;
  setMaxTurns(3) |> ping;

  setPing(ping) |> pong;

  startMsg() |> ping;
}