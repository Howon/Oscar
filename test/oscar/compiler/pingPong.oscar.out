ping
  pong
ping
  pong
ping
  pong
ping
ping stopped
pong stopped
